File: src/com/OpenSerenity/elements/BasePage.java
package com.OpenSerenity.elements;

import com.OpenSerenity.core.Browser;
import com.OpenSerenity.core.TestContext;
import com.OpenSerenity.functionalInterfaces.Action;
import com.OpenSerenity.functionalInterfaces.Func;
import com.OpenSerenity.functionalInterfaces.WaitCondition;
import com.OpenSerenity.utils.Waiter;

public class BasePage {
    protected BaseElement content;
    protected String windowHandle;

    public static <TPage extends BasePage> TPage open(Class<TPage> clz, Action action)
            throws Exception {
        TPage page = clz.newInstance();
        page.open(action);
        return page;
    }

    public static <TPage extends BasePage> TPage openWindow(Class<TPage> clz, Action action)
            throws IllegalAccessException, InstantiationException {
        TPage page = clz.newInstance();
        page.openWindow(action);
        return page;
    }

    protected void open(Action action) throws Exception {
        action.invoke(null);
        windowHandle = TestContext.browser.getCurrentWindowHandle();
        init();
        waitForLoad();
    }

    protected void openWindow(Action action)
    {
        throw new UnsupportedOperationException();
    }

    protected void init() throws Exception {
        content = TestContext.browser.findElement(BaseElement.class, "body", new Func<Browser>() {
            @Override
            public Browser invoke() {
                TestContext.browser.selectWindow(windowHandle);
                TestContext.browser.selectFrame(null);
                return null;
            }
        });
    }

    public boolean isStale() throws Exception {
        return isClosed();
    }

    public String getTitle() {
            selectWindow();
            return TestContext.browser.getTitle();
    }

    public void waitForLoad() throws Exception {
        Waiter.Default().waitFor(new WaitCondition() {
            @Override
            public boolean invoke() {
                try {
                    return !isStale();
                } catch (Exception ex) {
                    return false;
                }

            }
        });
    }

    public void selectWindow()
    {
        try
        {
            if (windowHandle == TestContext.browser.getCurrentWindowHandle())
            {
                return;
            }
        }
        catch (Exception ex)
        {
            ex.printStackTrace();
        }
        TestContext.browser.selectWindow(windowHandle);
    }

    public void close()
    {
        if (isClosed())
        {
            return;
        }
        selectWindow();
        TestContext.browser.closeWindow(windowHandle);
    }

    public boolean isClosed() {
        return !TestContext.browser.getAllWindowsHandles().contains(windowHandle);
    }

    public BasePage acceptAlert()
    {
        TestContext.browser.acceptMessageBox();
        return this;
    }

    public BasePage acceptConfirmation()
    {
        TestContext.browser.acceptMessageBox();
        return this;
    }

    public BasePage dismissConfirmation()
    {
        TestContext.browser.dismissMessageBox();
        return this;
    }
}

File: src/scala/com/OpenSerenity/elements/BasePage.scala
package com.OpenSerenity.elements

import com.OpenSerenity.core.TestContext
import scala.Predef.String
import com.OpenSerenity.utils.Waiter

object BasePage {
  def open[TPage <: BasePage[TPage]](action: () => Unit)(implicit m: scala.reflect.Manifest[TPage]) = {
    val page: TPage = m.runtimeClass.newInstance().asInstanceOf
    page.open(action)
    page
  }
}

class BasePage[T <: BasePage[T]] {
  def open(action: () => Unit) = {
    action()
    windowHandle = TestContext.browser.getCurrentWindowHandle
    init
    waitForLoad
  }

  protected def openWindow(action: () => Unit) = throw new UnsupportedOperationException

  protected def init= {
    content = TestContext.browser.findElement[Element]( "body", () => {
        TestContext.browser.selectWindow(windowHandle)
        TestContext.browser.selectFrame(null)
        null
      })
  }

  def isStale = isClosed

  def getTitle: String = {
    selectWindow
    return TestContext.browser.getTitle
  }

  def waitForLoad = Waiter.default.waitFor(() => {
    try !isStale
    catch {
      case ex: Exception => false
    }
  })

  def selectWindow {
    try {
      if (windowHandle eq TestContext.browser.getCurrentWindowHandle) {
        return
      }
    }
    catch {
      case ex: Exception => {
        ex.printStackTrace
      }
    }
    TestContext.browser selectWindow windowHandle
  }

  def close: Unit = {
    if (isClosed) return
    selectWindow
    TestContext.browser closeWindow windowHandle
  }

  def isClosed: Boolean = {
    return !TestContext.browser.getAllWindowsHandles.contains(windowHandle)
  }

  def acceptAlert: BasePage[T] = {
    TestContext.browser.acceptMessageBox
    this
  }

  def acceptConfirmation: BasePage[T] = {
    TestContext.browser.acceptMessageBox
    this
  }

  def dismissConfirmation: BasePage[T] = {
    TestContext.browser.dismissMessageBox
    this
  }

  protected var content: BaseElement[_] = null
  protected var windowHandle: String = null
}


